File: CNigmaPlusPlus/Rotor.cpp
#include "pch.h"
#include "Rotor.h"

Rotor::Rotor(const ERotors& chosen_rotor) 
{
	switch (chosen_rotor)
	{
	case I: SetRotor("EKMFLGDQVZNTOWYHXUSPAIBRCJ", 'Q'); break;
	case II: SetRotor("AJDKSIRUXBLHWTMCQGZNPYFVOE", 'E'); break;
	case III: SetRotor("BDFHJLCPRTXVZNYEIWGAKMUSQO", 'V'); break;
	case IV: SetRotor("ESOVPZJAYQUIRHXLNFTGKDCMWB", 'J'); break;
	case V: SetRotor("VZBRGITYUPSDNHLXAWMJQOFECK", 'Z'); break;
	}

}

void Rotor::SetRotor(const string& wiring, const char& ring_notch)
{
	notch = ring_notch;
	right_rotor_wiring = wiring;
}


size_t Rotor::Forward(const size_t& in_signal)
{
	const char letter = right_rotor_wiring[in_signal];
	const size_t out_signal = left_rotor_letters.find(letter);

	return out_signal;
}

size_t Rotor::Backward(const size_t& in_signal)
{
	const char letter = left_rotor_letters[in_signal];
	const size_t out_signal = right_rotor_wiring.find(letter);

	return out_signal;
}

void Rotor::ShowRotor() const
{
	cout << right_rotor_wiring << "\n";
	cout << left_rotor_letters << "\n";
}

char Rotor::GetCurrentRotorLetter() const
{
	 return left_rotor_letters[0];
}

void Rotor::Rotate(const bool& forward, const size_t& num_rotations)
{

	if (forward)
	{
		for (size_t dex = num_rotations; dex != 0; dex--)
		{
			right_rotor_wiring = right_rotor_wiring.substr(1) + right_rotor_wiring[0];
			left_rotor_letters = left_rotor_letters.substr(1) + left_rotor_letters[0];
		}
	}
	else
	{
		for (size_t dex = num_rotations; dex != 0; dex--)
		{
			
			right_rotor_wiring = right_rotor_wiring[25] + right_rotor_wiring;
			left_rotor_letters = left_rotor_letters[25] + left_rotor_letters;
			right_rotor_wiring.pop_back();
			left_rotor_letters.pop_back();
		}
	}
	
}

void Rotor::RotateToLetter(const char& letter)
{
	const size_t n = left_rotor_letters.find(letter);
	Rotate(n);
}

File: PyNigma/rotor.py
from components import e_rotors


class Rotor:
    """
    The rotors are a key element to the Enigma Machine. This is based on the 1949 Enigma 2, which had 5 rotors.
    These are defined here and can be selected:

    I = EKMFLGDQVZNTOWYHXUSPAIBRCJ, Q
    II = AJDKSIRUXBLHWTMCQGZNPYFVOE, E
    III = BDFHJLCPRTXVZNYEIWGAKMUSQO, V
    IV = ESOVPZJAYQUIRHXLNFTGKDCMWB, J
    V = VZBRGITYUPSDNHLXAWMJQOFECK, Z
    """

    def __init__(self, wiring: e_rotors) -> None:
        self.left = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

        match wiring:
            case e_rotors.I:
                self.right = e_rotors.value[0]
                self.notch = e_rotors.value[1]
            case e_rotors.II:
                self.right = e_rotors.value[0]
                self.notch = e_rotors.value[1]
            case e_rotors.III:
                self.right = e_rotors.value[0]
                self.notch = e_rotors.value[1]
            case e_rotors.IV:
                self.right = e_rotors.value[0]
                self.notch = e_rotors.value[1]
            case e_rotors.V:
                self.right = e_rotors.value[0]
                self.notch = e_rotors.value[1]

    def forward(self, signal: int) -> int:
        letter = self.right[signal]
        signal = self.left.find(letter)
        return signal

    def backward(self, signal: int) -> int:
        letter = self.left[signal]
        signal = self.right.find(letter)
        return signal

    def show(self) -> None:
        print(self.left)
        print(self.right)
        print("")

    def rotate(self, n: int = 1, forward: bool = True):
        for i in range(n):
            if forward:
                self.left = self.left[1:] + self.left[0]
                self.right = self.right[1:] + self.right[0]
            else:
                self.left = self.left[25] + self.left[:25]
                self.right = self.right[25] + self.right[:25]

    def rotate_to_letter(self, letter: str):
        n = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".find(letter)
        self.rotate(n)

    def set_ring(self, n: int):

        # rotate the rotor backwards
        self.rotate(n - 1, forward=False)

        # adjust the turnover notch in relationship to the wiring
        n_notch = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".find(self.notch)
        self.notch = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"[(n_notch - (n - 1)) % 26]
